server/api.js

```diff
@@ -0,0 +1,409 @@
+]
+
+async function handleCreateClientGateway(req, res) {
+  const gateway = await db.createGateway(
+    req.body.gatewayId,
+    req.params.clientId,
+    req.body.displayName,
+    null,
+    req.body.isDisplayed,
+    req.body.isSendingVitals,
+  )
+
+  // Should the database query fail, db.createGateway should internally handle thrown errors and return either null or undefined.
+  // The status code 404 is used here as the failure was probably caused by the client not existing.
+  if (gateway == null) {
+    res.status(404).send({ status: 'error', message: 'Not Found' })
+
+    return
+  }
+
+  res.set('Location', `${req.path}/${gateway.id}`) // location of newly created gateway
+  res.status(201).send({ status: 'success', data: gateway })
+}
+
+const validateGetClient = Validator.param(['clientId']).notEmpty()
+
+async function handleGetClient(req, res) {
+  const client = await db.getClientWithId(req.params.clientId)
+
+  // Couldn't get the client; Not found.
+  if (client == null) {
+    res.status(404).send({ status: 'error', message: 'Not Found' })
+
+    return
+  }
+
+  res.status(200).send({ status: 'success', data: client })
+}
+
+async function handleGetClients(req, res) {
+  const clients = await db.getClients()
+
+  res.status(200).send({ status: 'success', data: clients })
+}
+
+const validateGetClientButton = Validator.param(['clientId', 'buttonId']).notEmpty()
+
+async function handleGetClientButton(req, res) {
+  const button = await db.getButtonWithId(req.params.buttonId)
+
+  // Couldn't get the button; Not found.
+  if (button == null || button.client.id !== req.params.clientId) {
+    res.status(404).send({ status: 'error', message: 'Not Found' })
+
+    return
+  }
+
+  res.status(200).send({ status: 'success', data: button })
+}
+
+const validateGetClientButtons = Validator.param(['clientId']).notEmpty()
+
+async function handleGetClientButtons(req, res) {
+  const buttons = await db.getButtonsWithClientId(req.params.clientId)
+
+  // if the query failed and returned null, the clientId is probably wrong
+  if (buttons == null) {
+    res.status(404).send({ status: 'error', message: 'Not Found' })
+
+    return
+  }
+
+  // something like this: buttons = buttons.map(button => { button.sdfs ... })
+  // (remove single field)
+
+  res.status(200).send({ status: 'success', data: buttons })
+}
+
+const validateGetClientButtonSessions = Validator.param(['clientId', 'buttonId']).notEmpty()
+
+async function handleGetClientButtonSessions(req, res) {
+  const button = await db.getButtonWithId(req.params.buttonId)
+
+  // check that this button exists and is owned by the specified client
+  // NOTE: if clientId is invalid, then the query will fail and return null
+  if (button == null || button.client.id !== req.params.clientId) {
+    res.status(404).send({ status: 'error', message: 'Not Found' })
+
+    return
+  }
+
+  const sessions = await db.getRecentSessionsWithButtonId(req.params.buttonId)
```

Comment:
rework - how many sessions does this return? Maybe add new parameter with default value, and SQL `LIMIT`?